File: packages/py/model/src/dbx_tools/model/classify.py
from __future__ import annotations

import math
from collections.abc import Iterable

from .models import (
    EndpointCapabilities,
    ModelClass,
    ModelFamily,
    ServingEndpointSummary,
    parse_model_name,
    version_tuple,
)
from .reasoning import reasoning_efforts_by_family

CHAT_TASK = "llm/v1/chat"
EMBEDDING_TASK = "llm/v1/embeddings"


def supports_tools_by_family(name: str) -> bool:
    parsed = parse_model_name(name)
    if parsed is None:
        return False
    if parsed.family == ModelFamily.GEMINI:
        return False
    if parsed.family == ModelFamily.GPT and "oss" in parsed.model:
        return False
    return parsed.family in {
        ModelFamily.CLAUDE,
        ModelFamily.GLM,
        ModelFamily.GPT,
        ModelFamily.LLAMA,
        ModelFamily.QWEN,
    }


def endpoint_capabilities(
    endpoint: ServingEndpointSummary | dict[str, object],
) -> EndpointCapabilities:
    summary = _summary(endpoint)
    embedding = summary.task == EMBEDDING_TASK or summary.model_class == ModelClass.EMBEDDING
    chat = not embedding and (summary.task == CHAT_TASK or summary.model_class is not None)
    tools = chat and (
        summary.supports_tools
        if summary.supports_tools is not None
        else supports_tools_by_family(summary.name)
    )
    reasoning_efforts = summary.reasoning_efforts or reasoning_efforts_by_family(summary.name)
    return EndpointCapabilities(
        chat=chat,
        embedding=embedding,
        tools=tools,
        reasoningEfforts=reasoning_efforts,
    )


def classify_by_family(name: str) -> dict[str, object] | None:
    parsed = parse_model_name(name)
    if parsed is None:
        return None
    model = set(parsed.model)

    def result(model_class: ModelClass) -> dict[str, object]:
        major, minor, patch = version_tuple(name)
        return {"class": model_class.value, "rank": major * 1_000_000 + minor * 1_000 + patch}

    if parsed.family == ModelFamily.CLAUDE and "opus" in model:
        return result(ModelClass.CHAT_THINKING)
    if parsed.family == ModelFamily.CLAUDE and "sonnet" in model:
        return result(ModelClass.CHAT_BALANCED)
    if parsed.family == ModelFamily.CLAUDE and "haiku" in model:
        return result(ModelClass.CHAT_FAST)
    if parsed.family == ModelFamily.GPT and "oss" in model:
        return result(ModelClass.CHAT_BALANCED if "120b" in model else ModelClass.CHAT_FAST)
    if parsed.family == ModelFamily.GPT:
        if "pro" in model:
            return result(ModelClass.CHAT_THINKING)
        if "mini" in model or "nano" in model:
            return result(ModelClass.CHAT_FAST)
        return result(ModelClass.CHAT_BALANCED)
    if parsed.family == ModelFamily.GEMINI:
        if {"flash", "lite"} <= model:
            return result(ModelClass.CHAT_FAST)
        if "pro" in model:
            return result(ModelClass.CHAT_THINKING)
        return result(ModelClass.CHAT_BALANCED)
    if parsed.family == ModelFamily.GEMMA:
        return result(ModelClass.CHAT_FAST)
    if parsed.family == ModelFamily.LLAMA:
        if "maverick" in model or "405b" in model:
            return result(ModelClass.CHAT_THINKING)
        if "70b" in model:
            return result(ModelClass.CHAT_BALANCED)
        if "8b" in model or "1b" in model:
            return result(ModelClass.CHAT_FAST)
        return result(ModelClass.CHAT_BALANCED)
    if parsed.family == ModelFamily.QWEN:
        return result(ModelClass.CHAT_BALANCED)
    return None


def classify_endpoints(
    endpoints: Iterable[ServingEndpointSummary | dict[str, object]],
) -> dict[str, list[dict[str, object]]]:
    summaries = [_summary(endpoint) for endpoint in endpoints]
    chat = [endpoint for endpoint in summaries if endpoint.task == CHAT_TASK]
    qualities = sorted(
        endpoint.profile.quality
        for endpoint in chat
        if endpoint.profile is not None
        and endpoint.profile.quality is not None
        and math.isfinite(endpoint.profile.quality)
    )
    low = _quantile(qualities, 1 / 3)
    high = _quantile(qualities, 2 / 3)
    buckets: dict[ModelClass, list[tuple[tuple[object, ...], ServingEndpointSummary]]] = {
        model_class: [] for model_class in ModelClass
    }

    for endpoint in chat:
        quality = endpoint.profile.quality if endpoint.profile else None
        if quality is not None and math.isfinite(quality):
            model_class = (
                ModelClass.CHAT_THINKING
                if quality >= high
                else ModelClass.CHAT_FAST
                if quality <= low
                else ModelClass.CHAT_BALANCED
            )
            cost = (
                endpoint.profile.cost
                if endpoint.profile and endpoint.profile.cost is not None
                else math.inf
            )
            speed = (
                endpoint.profile.speed
                if endpoint.profile and endpoint.profile.speed is not None
                else 0
            )
            key = (0, -quality, cost, -speed, *[-part for part in version_tuple(endpoint.name)])
        else:
            family = classify_by_family(endpoint.name)
            if family is None:
                continue
            model_class = ModelClass(str(family["class"]))
            key = (
                1,
                -int(family["rank"]),
                math.inf,
                0,
                *[-part for part in version_tuple(endpoint.name)],
            )
        buckets[model_class].append((key, endpoint))

    embeddings = [endpoint for endpoint in summaries if endpoint.task == EMBEDDING_TASK]
    return {
        ModelClass.CHAT_THINKING.value: _dump_sorted(buckets[ModelClass.CHAT_THINKING]),
        ModelClass.CHAT_BALANCED.value: _dump_sorted(buckets[ModelClass.CHAT_BALANCED]),
        ModelClass.CHAT_FAST.value: _dump_sorted(buckets[ModelClass.CHAT_FAST]),
        ModelClass.EMBEDDING.value: [endpoint.as_dict() for endpoint in embeddings],
    }


def classified_summaries(
    endpoints: Iterable[ServingEndpointSummary | dict[str, object]],
) -> dict[ModelClass, list[ServingEndpointSummary]]:
    return {
        ModelClass(key): [ServingEndpointSummary.model_validate(value) for value in values]
        for key, values in classify_endpoints(endpoints).items()
    }


def _summary(value: ServingEndpointSummary | dict[str, object]) -> ServingEndpointSummary:
    return (
        value
        if isinstance(value, ServingEndpointSummary)
        else ServingEndpointSummary.model_validate(value)
    )


def _quantile(values: list[float], probability: float) -> float:
    if not values:
        return math.nan
    index = (len(values) - 1) * probability
    lower = math.floor(index)
    upper = math.ceil(index)
    if lower == upper:
        return values[lower]
    return values[lower] + (values[upper] - values[lower]) * (index - lower)


def _dump_sorted(
    values: list[tuple[tuple[object, ...], ServingEndpointSummary]],
) -> list[dict[str, object]]:
    return [endpoint.as_dict() for _, endpoint in sorted(values, key=lambda value: value[0])]

File: packages/rs/model/src/classify.rs
//! Model capability classification and stable class ordering.

use std::cmp::Ordering;

use crate::models::{
    parse_model_name, version_tuple, ModelClass, ModelFamily, ServingEndpointSummary,
};

pub const CHAT_TASK: &str = "llm/v1/chat";
pub const EMBEDDING_TASK: &str = "llm/v1/embeddings";

pub fn supports_tools_by_family(name: &str) -> bool {
    let Some(parsed) = parse_model_name(name) else {
        return false;
    };
    if parsed.family == ModelFamily::Gemini
        || (parsed.family == ModelFamily::Gpt && parsed.model.iter().any(|part| part == "oss"))
    {
        return false;
    }
    matches!(
        parsed.family,
        ModelFamily::Claude
            | ModelFamily::Glm
            | ModelFamily::Gpt
            | ModelFamily::Llama
            | ModelFamily::Qwen
    )
}

pub fn classify_by_family(name: &str) -> Option<(ModelClass, u64)> {
    let parsed = parse_model_name(name)?;
    let version = version_tuple(name);
    let rank =
        u64::from(version[0]) * 1_000_000 + u64::from(version[1]) * 1_000 + u64::from(version[2]);
    let has = |part: &str| parsed.model.iter().any(|candidate| candidate == part);
    let model_class = match parsed.family {
        ModelFamily::Claude if has("opus") => ModelClass::ChatThinking,
        ModelFamily::Claude if has("sonnet") => ModelClass::ChatBalanced,
        ModelFamily::Claude if has("haiku") => ModelClass::ChatFast,
        ModelFamily::Gpt if has("oss") && has("120b") => ModelClass::ChatBalanced,
        ModelFamily::Gpt if has("oss") => ModelClass::ChatFast,
        ModelFamily::Gpt if has("pro") => ModelClass::ChatThinking,
        ModelFamily::Gpt if has("mini") || has("nano") => ModelClass::ChatFast,
        ModelFamily::Gpt => ModelClass::ChatBalanced,
        ModelFamily::Gemini if has("flash") && has("lite") => ModelClass::ChatFast,
        ModelFamily::Gemini if has("pro") => ModelClass::ChatThinking,
        ModelFamily::Gemini => ModelClass::ChatBalanced,
        ModelFamily::Gemma => ModelClass::ChatFast,
        ModelFamily::Llama if has("maverick") || has("405b") => ModelClass::ChatThinking,
        ModelFamily::Llama if has("8b") || has("1b") => ModelClass::ChatFast,
        ModelFamily::Llama => ModelClass::ChatBalanced,
        ModelFamily::Qwen => ModelClass::ChatBalanced,
        _ => return None,
    };
    Some((model_class, rank))
}

pub fn classify_endpoints(
    endpoints: &[ServingEndpointSummary],
) -> Vec<(ModelClass, ServingEndpointSummary)> {
    let mut qualities = endpoints
        .iter()
        .filter(|endpoint| endpoint.task.as_deref() == Some(CHAT_TASK))
        .filter_map(|endpoint| endpoint.profile.as_ref()?.quality)
        .filter(|quality| quality.is_finite())
        .collect::<Vec<_>>();
    qualities.sort_by(f64::total_cmp);
    let low = quantile(&qualities, 1.0 / 3.0);
    let high = quantile(&qualities, 2.0 / 3.0);

    let mut classified = endpoints
        .iter()
        .filter_map(|endpoint| {
            if endpoint.task.as_deref() == Some(EMBEDDING_TASK) {
                return Some((
                    ModelClass::Embedding,
                    SortKey::embedding(),
                    endpoint.clone(),
                ));
            }
            if endpoint.task.as_deref() != Some(CHAT_TASK) {
                return None;
            }
            let quality = endpoint
                .profile
                .as_ref()
                .and_then(|profile| profile.quality)
                .filter(|value| value.is_finite());
            if let Some(quality) = quality {
                let model_class = if quality >= high {
                    ModelClass::ChatThinking
                } else if quality <= low {
                    ModelClass::ChatFast
                } else {
                    ModelClass::ChatBalanced
                };
                let profile = endpoint.profile.as_ref().expect("profile has quality");
                return Some((
                    model_class,
                    SortKey {
                        source: 0,
                        rank: quality,
                        cost: profile.cost.unwrap_or(f64::INFINITY),
                        speed: profile.speed.unwrap_or(0.0),
                        version: version_tuple(&endpoint.name),
                    },
                    endpoint.clone(),
                ));
            }
            let (model_class, rank) = classify_by_family(&endpoint.name)?;
            Some((
                model_class,
                SortKey {
                    source: 1,
                    rank: rank as f64,
                    cost: f64::INFINITY,
                    speed: 0.0,
                    version: version_tuple(&endpoint.name),
                },
                endpoint.clone(),
            ))
        })
        .collect::<Vec<_>>();
    classified.sort_by(|left, right| {
        left.0
            .order()
            .cmp(&right.0.order())
            .then_with(|| left.1.compare(&right.1))
    });
    classified
        .into_iter()
        .map(|(model_class, _, endpoint)| (model_class, endpoint))
        .collect()
}

#[derive(Clone, Copy)]
struct SortKey {
    source: u8,
    rank: f64,
    cost: f64,
    speed: f64,
    version: [u32; 3],
}

impl SortKey {
    fn embedding() -> Self {
        Self {
            source: 0,
            rank: 0.0,
            cost: 0.0,
            speed: 0.0,
            version: [0, 0, 0],
        }
    }

    fn compare(&self, other: &Self) -> Ordering {
        self.source
            .cmp(&other.source)
            .then_with(|| other.rank.total_cmp(&self.rank))
            .then_with(|| self.cost.total_cmp(&other.cost))
            .then_with(|| other.speed.total_cmp(&self.speed))
            .then_with(|| other.version.cmp(&self.version))
    }
}

fn quantile(values: &[f64], probability: f64) -> f64 {
    if values.is_empty() {
        return f64::NAN;
    }
    let index = (values.len() - 1) as f64 * probability;
    let lower = index.floor() as usize;
    let upper = index.ceil() as usize;
    if lower == upper {
        values[lower]
    } else {
        values[lower] + (values[upper] - values[lower]) * (index - lower as f64)
    }
}
